cpp/arcticdb/version/version_map.hpp

```diff
@@ -580,68 +579,40 @@ class VersionMapImpl {
-                if (requested_load_type == LoadType::LOAD_LATEST) {
-                    return true;
-                }
-                if (requested_load_type == LoadType::LOAD_DOWNTO) {
-                    return loaded_as_far_as_load_until(*entry, load_param);
-                }
-                break;
-            case LoadType::LOAD_LATEST_UNDELETED:
-                if (requested_load_type == LoadType::LOAD_LATEST_UNDELETED
-                    || requested_load_type == LoadType::LOAD_LATEST) {
-                    return true;
-                }
-
-                if (requested_load_type == LoadType::LOAD_DOWNTO) {
-                    return loaded_as_far_as_load_until(*entry, load_param);
-                }
-                break;
+                return true;
+            case LoadType::LOAD_LATEST: {
+                // If entry has at least one (maybe undeleted) index we have the latest value cached
+
+                // This check can be slow if we have thousands of deleted versions before the first undeleted and we're
+                // looking for an undeleted version. If that is ever a problem we can just store a boolean whether
+                // we have an undeleted version.
+                auto opt_latest = entry->get_first_index(requested_load_strategy.include_deleted_).first;
+                return opt_latest.has_value();
+            }
             case LoadType::LOAD_DOWNTO:
-                if (requested_load_type == LoadType::LOAD_DOWNTO) {
-                    return loaded_as_far_as_load_until(*entry, load_param);
-                }
-
-                if (requested_load_type == LoadType::LOAD_LATEST_UNDELETED) {
-                    auto opt_latest = entry->get_first_index(false).first;
-                    return opt_latest.has_value();
-                }
-
-                if (requested_load_type == LoadType::LOAD_LATEST) {
-                    auto opt_latest = entry->get_first_index(true).first;
-                    return opt_latest.has_value();
+                // We check whether the oldest loaded version is before or at the requested one
+                return loaded_as_far_as_version_id(*entry, requested_load_strategy.load_until_version_.value());
+            case LoadType::LOAD_FROM_TIME: {
+                // We check whether the cached (deleted or undeleted) timestamp is before or at the requested one
+                auto cached_timestamp = requested_load_strategy.include_deleted_ ?
+                                        entry->loaded_with_progress_.earliest_loaded_timestamp_ :
+                                        entry->loaded_with_progress_.earliest_loaded_undeleted_timestamp_;
+                return cached_timestamp <= requested_load_strategy.load_from_time_.value();
+            }
+            case LoadType::LOAD_ALL:
+                // We can have all versions cached only when cache was loaded by loading all versions.
```

Comment:
I don't think this is strictly true, as you can load all of the information in the version chain without the load parameter having been `LOAD_ALL`
e.g. loading down to a timestamp before version 0 was written
Probably not worth worrying about here as it is just a cache, but worth correcting the comment.

Reply:
Agreed, updated the comment.